sretoolbox/utils/multiprocess.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2021 Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Multiprocessing abstractions.
+"""
+
+import multiprocessing
+from multiprocessing import Pool
+
+
+def run(func, iterable, process_pool_size,
+        return_exceptions=False, **kwargs):
+    """run_processes executes a function for each item in the input iterable.
+    execution will be done in a processpool according to the input
+    process_pool_size.  kwargs are passed to the input function
+    (optional). If return_exceptions is true, any exceptions that may
+    have happened in each thread are returned in the return value,
+    allowing the caller to get as much work done as possible.
+    """
+    mpm = multiprocessing.Manager()
+
+    process_list = [(func, [i], kwargs, mpm.dict()) for i in iterable]
+    with Pool(process_pool_size) as pool:
+        pool.map(_process_task, process_list)
+
+    results = []
+    for process_item in process_list:
+        if "success" in process_item[3]:
+            results.append(process_item[3]["success"])
+        elif "exception" in process_item[3] and return_exceptions:
+            results.append(process_item[3]["exception"])
+        elif "exception" in process_item[3]:
+            raise process_item[3]["exception"]
+    return results
+
+
+def _process_task(spec):
+    func, args, kwargs, result = spec
+    try:
+        result["success"] = func(*args, **kwargs)
+    # pylint: disable=broad-except
+    except Exception as details:
+        result["exception"] = details
+    except SystemExit as details:
+        if int(details.code) != 0:
+            result["exception"] = details
+        result["success"] = None
```

Comment:
This api is a bit misleading... "success" with value None is actually a good thing? Maybe "success" should be a Boolean that always exist, and based on its value, you look for the result or for the exception.

Reply:
I would like it more too, I agree with @apahim's comment.

Reply:
Ok, so, let me dig a bit deeper here...

In your `_process_task()`, you'e writing to a shared unique dictionary, created by calling `mpm.dict()`. The fact the ditionary is both shared and unique defeats it's purpose. I know you're sharing each unique dict between the caller and process, but hey, `return` exists for that. A shared dict is usefull to share data between processes. So, you should be actually using a `return` in this `_process_task()` function.

That said, the returned structure can still be a dictionary, and my previous comment asks you to use `success = True/False` as the signal to the caller, instead of the "success"/"exception" presence/absense.

To sum it up, this function should probably look lile this:

```python
def _process_task(spec):
    func, args, kwargs = spec
    task_result = dict()
    try:
        task_result["return"] = func(*args, **kwargs)
        task_result["success"] = True
        task_result["exception"] = None
    # pylint: disable=broad-except
    except Exception as details:
        task_result["return"] = None
        task_result["success"] = False
        task_result["exception"] = details
    except SystemExit as details:
        if int(details.code) == 0:
            task_result["return"] = None
            task_result["success"] = True
            task_result["exception"] = None
        else:
            task_result["return"] = None
            task_result["success"] = False
            task_result["exception"] = details

    return task_result
```

Please notice there's a `return` and the `task_result` is a local dict.

With that, the caller would need some refactoring as well...

```python
    ...
    # No need for a mpm.dict() here, as you're not really sharing the dict between processes
    process_list = [(func, [i], kwargs) for i in iterable]
    with Pool(process_pool_size) as pool:
        # pool.map() creates a list with the returns from the processes
        tasks_results = pool.map(_process_task, process_list)
    ...
```

Then you'd have to process `tasks_results`, instead of carving the `mpm.dict()`s you were sharing between the caller and the calees. Something like this

```python
    ...
    run_results = []
    for tr in tasks_results:
        if tr["success"]:
            run_results.append(tr["return"])
            continue
        if return_exceptions:
            run_results.append(tr["exception"])
            continue
        raise tr["exception"]
    return run_results
```

Reply:
Again, the way you did it works nicelly, it's just a bit off semantically, so no big deal :)

Reply:
oh, ok, this looks really nice! I didn't want to change the essence of your logic, just the semantics, but I'm glad that you saw it!